File: app/schemas/email.py
"""
The canonical parsed-email object and the relay hop model.

M1 (ingestion) is the only thing that builds a ParsedEmail. Everything
downstream reads it and never re-parses raw bytes -- except M3, which needs
the original bytes for DKIM verification, so we keep them on the object.

Owner: Track A
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class HopTrust(str, Enum):
    """
    Where a hop sits relative to the trust boundary.

    See the architecture doc, section 03. Received headers are prepended, so
    the chain reads bottom-to-top in transmission order, and an attacker can
    forge every hop below the first server we control. We walk down from the
    top through hosts we can authenticate and stop at the last one we trust:
    the IP that hop received *from* is the highest-confidence origin.
    """

    TRUSTED = "TRUSTED"        # our infrastructure or an authenticated provider
    BOUNDARY = "BOUNDARY"      # the earliest trustworthy hop -- report this origin
    UNVERIFIED = "UNVERIFIED"  # below the boundary; attacker could have written it


class Hop(BaseModel):
    """One Received: header, parsed."""

    seq: int = Field(..., description="0 = bottom of the header block = claimed origin.")
    raw: str = Field(..., description="The original header line, unmodified.")

    from_host: str | None = None
    from_ip: str | None = None
    by_host: str | None = None
    by_ip: str | None = None
    protocol: str | None = Field(None, description="e.g. ESMTPS, ESMTPSA, SMTP")
    timestamp: datetime | None = None

    trust: HopTrust = HopTrust.UNVERIFIED
    anomalies: list[str] = Field(
        default_factory=list,
        description="Signal keys raised against this specific hop, e.g. "
        "['private_ip_in_public_chain', 'rdns_mismatch'].",
    )

    # Filled by M5 (Track D). Left empty when intel is unavailable.
    geo: dict[str, Any] | None = None


class Attachment(BaseModel):
    filename: str | None = None
    content_type: str | None = None
    size_bytes: int = 0
    sha256: str


class ExtractedUrl(BaseModel):
    url: str
    domain: str | None = None
    display_text: str | None = None
    is_shortened: bool = False
    mismatched_anchor: bool = Field(
        False,
        description="True when the visible link text names a different domain "
        "than the href. A classic phishing tell.",
    )


class ParsedEmail(BaseModel):
    """
    The canonical internal representation. One ingestion path or ten, every
    analyzer sees only this.
    """

    sha256: str = Field(..., description="Hash of the original raw bytes. "
                                        "Goes in the forensic report as the evidence identifier.")
    raw_bytes: bytes = Field(..., repr=False, exclude=True)

    # ordered exactly as they appeared, duplicates preserved -- order is evidence
    headers: list[tuple[str, str]] = Field(default_factory=list)

    message_id: str | None = None
    subject: str | None = None
    date: datetime | None = None

    from_display_name: str | None = None
    from_addr: str | None = None
    reply_to: str | None = None
    return_path: str | None = None
    to_addrs: list[str] = Field(default_factory=list)

    body_text: str = ""
    body_html: str | None = None

    urls: list[ExtractedUrl] = Field(default_factory=list)
    attachments: list[Attachment] = Field(default_factory=list)

    def header_values(self, name: str) -> list[str]:
        """All values for a header name, in original order. Case-insensitive."""
        low = name.lower()
        return [v for k, v in self.headers if k.lower() == low]

    def first_header(self, name: str) -> str | None:
        vals = self.header_values(name)
        return vals[0] if vals else None

File: web/src/lib/schemas/email.ts
/**
 * The canonical parsed-email object and the relay-hop model
 * (ported from app/schemas/email.py).
 *
 * M1 (ingestion) is the only thing that builds a ParsedEmail. Everything
 * downstream reads it and never re-parses raw bytes -- except M3, which needs
 * the original bytes for DKIM verification, so we keep them on the object.
 */

/** Where a hop sits relative to the trust boundary. */
export enum HopTrust {
  TRUSTED = 'TRUSTED', // our infrastructure or an authenticated provider
  BOUNDARY = 'BOUNDARY', // the earliest trustworthy hop -- report this origin
  UNVERIFIED = 'UNVERIFIED', // below the boundary; attacker could have written it
}

/** One Received: header, parsed. */
export interface Hop {
  seq: number; // 0 = bottom of the header block = claimed origin
  raw: string;
  from_host: string | null;
  from_ip: string | null;
  by_host: string | null;
  by_ip: string | null;
  protocol: string | null;
  timestamp: string | null; // ISO
  trust: HopTrust;
  anomalies: string[];
  geo: Record<string, unknown> | null;
}

export interface Attachment {
  filename: string | null;
  content_type: string | null;
  size_bytes: number;
  sha256: string;
  /** Active content extracted from an SVG attachment (scripts / event handlers /
   *  javascript: or data: hrefs). Empty/absent for benign or non-SVG parts. */
  active_content?: string[];
}

export interface ExtractedUrl {
  url: string;
  domain: string | null;
  display_text: string | null;
  is_shortened: boolean;
  /** True when the visible link text names a different domain than the href. */
  mismatched_anchor: boolean;
}

export interface ParsedEmail {
  sha256: string;
  /** Original raw bytes; kept for M3 DKIM. Never serialized to the wire. */
  rawBytes: Buffer;
  /** Ordered exactly as they appeared, duplicates preserved -- order is evidence. */
  headers: Array<[string, string]>;
  message_id: string | null;
  subject: string | null;
  date: string | null; // ISO
  from_display_name: string | null;
  from_addr: string | null;
  reply_to: string | null;
  return_path: string | null;
  to_addrs: string[];
  body_text: string;
  body_html: string | null;
  urls: ExtractedUrl[];
  attachments: Attachment[];
}

/** All values for a header name, in original order. Case-insensitive. */
export function headerValues(email: ParsedEmail, name: string): string[] {
  const low = name.toLowerCase();
  return email.headers.filter(([k]) => k.toLowerCase() === low).map(([, v]) => v);
}

export function firstHeader(email: ParsedEmail, name: string): string | null {
  const vals = headerValues(email, name);
  return vals.length ? vals[0] : null;
}
